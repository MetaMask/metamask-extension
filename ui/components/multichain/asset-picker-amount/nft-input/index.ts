export { NFTInput } from './nft-input';
