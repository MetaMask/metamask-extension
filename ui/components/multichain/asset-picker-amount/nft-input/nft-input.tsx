import React from 'react';
import UnitInput from '../../../ui/unit-input';
import { Numeric } from '../../../../../shared/modules/Numeric';

type NFTInputProps = {
  integerValue: number;
  onChange?: (newAmountRaw: string, newAmountFormatted: string) => void;
  className?: string;
};

const NUMBERS_REGEX = /^[0-9]*$/u;

export function NFTInput({ integerValue, onChange, className }: NFTInputProps) {
  const handleChange = (newValueAsString: number) => {
    if (!onChange) {
      return;
    }

    const newValue = new Numeric(newValueAsString, 10);

    if (!Number.isInteger(newValue.toNumber())) {
      return;
    }

    onChange(newValue.toPrefixedHexString(), String(newValueAsString));
  };

  return (
    <UnitInput
      isDisabled={!onChange}
      isFocusOnInput={Boolean(onChange)}
      type="number"
      step={1}
      min={0}
      dataTestId="nft-input"
      onChange={handleChange}
      value={integerValue}
      className={className}
      keyPressRegex={NUMBERS_REGEX}
    />
  );
}
