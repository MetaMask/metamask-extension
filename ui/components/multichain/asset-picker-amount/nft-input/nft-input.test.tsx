import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import mockSendState from '../../../../../test/data/mock-send-state.json';
import configureStore from '../../../../store/store';
import { NFTInput } from './nft-input';

const createStore = ({ sendInputCurrencySwitched }: Record<string, boolean>) =>
  configureStore({
    ...mockSendState,
    metamask: {
      ...mockSendState.metamask,
    },
    appState: { ...mockSendState.appState, sendInputCurrencySwitched },
  });

describe('NFTInput', () => {
  it('matches snapshot', () => {
    const mockAssetChange = jest.fn();

    const { asFragment } = render(
      <Provider
        store={createStore({
          sendInputCurrencySwitched: true,
        })}
      >
        <NFTInput integerValue={1} onChange={() => mockAssetChange()} />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('disables input when change handler is not passed', () => {
    const { getByTestId } = render(
      <Provider
        store={createStore({
          sendInputCurrencySwitched: true,
        })}
      >
        <NFTInput integerValue={1} />
      </Provider>,
    );

    expect(getByTestId('nft-input')).toBeDisabled();
  });

  it('registers changes', () => {
    const mockAssetChange = jest.fn();

    const { queryByTestId } = render(
      <Provider
        store={createStore({
          sendInputCurrencySwitched: true,
        })}
      >
        <NFTInput integerValue={0} onChange={mockAssetChange} />
      </Provider>,
    );

    const input = queryByTestId('nft-input');
    expect(input).toBeInTheDocument();
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    fireEvent.change(input!, { target: { value: '10' } });

    expect(mockAssetChange).toHaveBeenCalledWith('0xa', '10');
  });
});
