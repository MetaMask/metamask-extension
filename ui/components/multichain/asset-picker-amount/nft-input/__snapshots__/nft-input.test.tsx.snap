// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTInput matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="unit-input"
  >
    <div
      class="unit-input__inputs"
    >
      <div
        class="unit-input__input-container"
        style="display: inherit;"
        title="1"
      >
        <input
          class="unit-input__input"
          data-testid="nft-input"
          dir="ltr"
          min="0"
          placeholder="0"
          step="any"
          style="width: 1.5ch;"
          type="number"
          value="1"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
